### PR TITLE
Upgrade blockchain-contracts to 0.3.1

### PR DESCRIPTION
**Pull request by D4nte**
Resolves comit-network/blockchain-contracts#40

### CODE REVIEW

**Comment by D4nte** [comment]
I challenge the people to improve my latest commit bcb6a46b, most likely by only using slices instead of `Vec`.

File: cnd/src/asset/ethereum/erc20.rs [hunk 1]
```diff
@@ -48,8 +48,23 @@ impl From<Erc20Quantity> for blockchain_contracts::ethereum::TokenQuantity {
...
+        if vec.len() > 32 {
+            unreachable!("There is already an `unreachable!` for that few lines above")
+        }
+        if vec.len() < 32 {
+            let mut full_len_vec = Vec::new();
+            let delta = 32 - vec.len();
+            full_len_vec.resize(delta, 0);
+            full_len_vec.extend_from_slice(&vec);
+            buf.copy_from_slice(&full_len_vec);
+        } else {
```
Comment (luckysori) [inline cnd/src/asset/ethereum/erc20.rs]: :question: Do you need this `else`?

Comment (D4nte) [reply cnd/src/asset/ethereum/erc20.rs]: Mini/easy-optimisation

**Comment by tcharding** [comment]
> I challenge the people to improve my latest commit [bcb6a46](https://github.com/comit-network/comit-rs/commit/bcb6a46b7526eaae449b64951a0f41ca27d57a8b), most likely by only using slices instead of `Vec`.

Challenge accepted :)

```
diff --git a/cnd/src/asset/ethereum/erc20.rs b/cnd/src/asset/ethereum/erc20.rs
index 48e4d55e..ff5711d9 100644
--- a/cnd/src/asset/ethereum/erc20.rs
+++ b/cnd/src/asset/ethereum/erc20.rs
@@ -41,6 +41,8 @@ impl FromWei<U256> for Erc20Quantity {
     }
 }

+use itertools::Itertools;
+
 impl From<Erc20Quantity> for blockchain_contracts::ethereum::TokenQuantity {
     fn from(quantity: Erc20Quantity) -> Self {
         if quantity > Erc20Quantity::max_value() {
@@ -52,18 +54,7 @@ impl From<Erc20Quantity> for blockchain_contracts::ethereum::TokenQuantity {
         let mut buf = [0u8; 32];
         let vec = quantity.0.to_bytes_be();

-        if vec.len() > 32 {
-            unreachable!("There is already an `unreachable!` for that few lines above")
-        }
-        if vec.len() < 32 {
-            let mut full_len_vec = Vec::new();
-            let delta = 32 - vec.len();
-            full_len_vec.resize(delta, 0);
-            full_len_vec.extend_from_slice(&vec);
-            buf.copy_from_slice(&full_len_vec);
-        } else {
-            buf.copy_from_slice(&vec);
-        }
+        buf.iter_mut().set_from(vec.iter().cloned());

         // TokenQuantity stores a Big Endian Byte Array
         blockchain_contracts::ethereum::TokenQuantity(buf)
```

**Comment by D4nte** [comment]
> use itertools::Itertools;

Ok, let's do it. I hope it's tested because from reading the doc I don't think it'll work.

**Comment by D4nte** [comment]
> > Challenge accepted :)
> 
> Ok, let's do it. I hope it's tested because from reading the doc I don't think it'll work.

Yep, as I expected, test fail @tcharding....

```
Expected :[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
Actual   :[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
<Click to see difference>

thread 'asset::ethereum::erc20::tests::blockchain_contracts_test::given_small_erc20_quantity_convert_to_token_quantity' panicked at 'assertion failed: `(left == right)`
  left: `[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]`,
 right: `[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]`', cnd/src/asset/ethereum/erc20.rs:263:13
```

If `vec.len()` is less than 32 then it needs to be left-padded with zeros.

**Comment by D4nte** [comment]
Why bors, why?

**Pull request merged**